meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java

```diff
@@ -0,0 +1,136 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.helix.metaclient.api.MetaClientInterface;
+import org.apache.helix.metaclient.datamodel.DataRecord;
+import org.apache.helix.metaclient.factories.MetaClientConfig;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+
+public class DistributedSemaphore {
+
+  /**
+   * Create a distributed semaphore client and creates a semaphore at the indicated path with the given capacity.
+   */
+  public DistributedSemaphore(MetaClientConfig config, String path, int capacity) {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Create a distributed semaphore client. Connects to an existing semaphore at given path.
+   */
+  public DistributedSemaphore(MetaClientConfig config, String path) {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Connect to an existing distributed semaphore client and semaphore at the indicated path.
+   */
+  public DistributedSemaphore(MetaClientInterface<DataRecord> client, String path) {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Acquire a permit. If no permit is available, log error and return null.
+   * @return a permit
+   */
+  public Permit acquire() {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+
+  /**
+   * Try to acquire multiple permits. If not enough permits are available, log error and return null.
+   * @param count number of permits to acquire
+   * @return a collection of permits
+   */
+  public Collection<Permit> acquire(int count) {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Try to acquire a permit. If no enough permit is available, wait for a specific time or return when it was able to acquire.
+   * If timeout <=0, then return immediately when not able to acquire.
+   * @param count number of permits to acquire
+   * @param timeout time to wait
+   * @param unit time unit
+   * @return a collection of permits
+   */
+  public Collection<Permit> acquire(int count, long timeout, TimeUnit unit) {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Get the remaining capacity of the semaphore
+   * @return remaining capacity
+   */
+  public long getRemainingCapacity() {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Get the semaphore data record
+   * @return semaphore data record
+   */
+  private DataRecord getSemaphore() {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Return a permit. If the permit is already returned, it will be ignored and throw a log info.
```

Comment:
throw an exception...?

Reply:
Mind my weird phrasing but i meant to say we can just log that the permit has already been returned. Since the end result is the same for success and failure, I don't think throwing an exception is necessary.